File: apps/runtime/src/git_repository/integration/cherry_pick.rs
use super::*;

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CherryPickPreview {
    pub target_oid: String,
    pub parents: Vec<String>,
    pub subject: String,
    pub author_name: String,
    pub author_email: String,
    pub author_time: String,
    pub mainline: Option<u32>,
    /// A merge commit requires an explicit parent selection before showing its
    /// relative patch or accepting a write. Nothing is inferred from HEAD.
    pub patch: Option<String>,
}
struct CommitData {
    parents: Vec<String>,
    subject: String,
    author_header: Vec<u8>,
    author_name: String,
    author_email: String,
    author_time: String,
}
impl RepositoryService {
    pub async fn cherry_pick_preview(
        &self,
        root: &Path,
        requested: &str,
        oid: &str,
        mainline: Option<u32>,
    ) -> AppResult<CherryPickPreview> {
        let normalized_oid = oid.to_ascii_lowercase();
        let oid = normalized_oid.as_str();
        require_oid(oid)?;
        let context = self.context(root, requested).await?;
        let token = Cancellation::default();
        let commit = self.pick_commit(&context, oid, &token).await?;
        let parent = pick_parent(&commit.parents, mainline, false)?;
        let patch = if commit.parents.len() > 1 && mainline.is_none() {
            None
        } else {
            Some(
                String::from_utf8_lossy(
                    &self
                        .read(&context.repository, pick_diff(oid, parent, true), &token)
                        .await?,
                )
                .into_owned(),
            )
        };
        Ok(CherryPickPreview {
            target_oid: oid.into(),
            parents: commit.parents,
            subject: commit.subject,
            author_name: commit.author_name,
            author_email: commit.author_email,
            author_time: commit.author_time,
            mainline,
            patch,
        })
    }

    pub(in super::super) async fn start_cherry_pick(
        &self,
        context: &RepositoryContext,
        action: &RepositoryAction,
        expected: &ExpectedState,
        operation: &Operation,
    ) -> AppResult<()> {
        let RepositoryAction::StartCherryPick {
            target_oid,
            mainline,
            record_origin,
            expected_state_token,
        } = action
        else {
            return Err(malformed());
        };
        let normalized_oid = target_oid.to_ascii_lowercase();
        let target_oid = &normalized_oid;
        let token = &operation.cancellation;
        let state = self.integration_snapshot(context, token).await?;
        if state.kind != "none"
            || state.head != *expected
            || state.state_token != *expected_state_token
        {
            return Err(AppError::Conflict(
                "Git state changed or an integration is already active; refresh and confirm again"
                    .into(),
            ));
        }
        if state.dirty
            || !state.conflicts.is_empty()
            || state.head.head_oid.is_none()
            || state.head.branch.is_none()
        {
            return Err(AppError::Conflict(
                "Cherry-pick requires a clean index and worktree on a committed local branch"
                    .into(),
            ));
        }
        let commit = self.pick_commit(context, target_oid, token).await?;
        let parent = pick_parent(&commit.parents, *mainline, true)?;
        let paths = self
            .read(
                &context.repository,
                pick_diff(target_oid, parent, false),
                token,
            )
            .await?;
        let touched = paths
            .split(|byte| *byte == 0)
            .filter(|path| !path.is_empty())
            .map(|path| {
                std::str::from_utf8(path)
                    .map(str::to_owned)
                    .map_err(|_| malformed())
            })
            .collect::<AppResult<std::collections::BTreeSet<_>>>()?;
        self.protect_local_paths(context, &touched, token).await?;
        let session_id = operation.snapshot.lock().expect("Git operation").id.clone();
        self.inner
            .integrations
            .lock()
            .expect("Git integrations")
            .insert(
                context.repository.clone(),
                IntegrationOwner {
                    session_id,
                    kind: "cherryPick",
                    mainline: *mainline,
                    original: expected.clone(),
                    target_oid: target_oid.clone(),
                    marker: None,
                },
            );
        let mut command = args(&[
            "-c",
            "core.editor=:",
            "-c",
            "rerere.enabled=false",
            "cherry-pick",
            "--no-rerere-autoupdate",
        ]);
        if let Some(mainline) = mainline {
            command.extend(["--mainline".into(), mainline.to_string()]);
        }
        if *record_origin {
            command.push("-x".into());
        }
        command.extend(["--".into(), target_oid.clone()]);
        let result = self
            .output(
                &context.repository,
                command,
                self.command_timeout,
                token,
                Some(operation.mutation_started.clone()),
            )
            .await;
        let actual = self
            .git_integration(context, &Cancellation::default())
            .await;
        let head = self
            .head(&context.repository, &Cancellation::default())
            .await;
        let mut owned = false;
        if let (Ok(actual), Ok(head)) = (&actual, &head) {
            let mut owners = self.inner.integrations.lock().expect("Git integrations");
            if let Some(owner) = owners.get_mut(&context.repository) {
                if actual.kind == "cherryPick"
                    && actual.target_oid.as_deref() == Some(target_oid)
                    && head == expected
                {
                    owner.marker = actual.marker.clone();
                    owned = owns_integration(owner, actual, head);
                } else if actual.kind == "none" {
                    owners.remove(&context.repository);
                }
            }
        }
        let output = result?;
        let actual = actual?;
        let head = head?;
        if owned && matches!(output.status, Some(0 | 1)) {
            operation.awaiting_resolution.store(true, Ordering::SeqCst);
            return Ok(());
        }
        if output.status != Some(0) {
            return Err(command_error(&output));
        }
        if actual.kind != "none" {
            return Err(AppError::Conflict(
                "Cherry-pick left an unverified Git sequence; inspect it before another action"
                    .into(),
            ));
        }
        self.verify_cherry_pick(context, target_oid, expected, &head, token)
            .await
    }

    pub(super) async fn verify_cherry_pick(
        &self,
        context: &RepositoryContext,
        target: &str,
        expected: &ExpectedState,
        head: &ExpectedState,
        token: &Cancellation,
    ) -> AppResult<()> {
        let original = expected.head_oid.as_deref().ok_or_else(malformed)?;
        let current = head.head_oid.as_deref().ok_or_else(malformed)?;
        let source = self.pick_commit(context, target, token).await?;
        let produced = self.pick_commit(context, current, token).await?;
        if head.branch != expected.branch
            || current == original
            || produced.parents != [original]
            || produced.author_header != source.author_header
        {
            return Err(AppError::Conflict("Resulting cherry-pick commit does not match its confirmed parent and source author; inspect HEAD before another action".into()));
        }
        Ok(())
    }

    async fn pick_commit(
        &self,
        context: &RepositoryContext,
        oid: &str,
        token: &Cancellation,
    ) -> AppResult<CommitData> {
        // Parse the actual commit header, not a history traversal, so shallow
        // boundary rewriting cannot misrepresent a non-root commit as a root.
        require_oid(oid)?;
        let kind = self
            .read(&context.repository, args(&["cat-file", "-t", oid]), token)
            .await?;
        if one_line(&kind)? != "commit" {
            return Err(AppError::BadRequest(
                "Cherry-pick target must be a commit object ID, not a tag or tree".into(),
            ));
        }
        let bytes = self
            .read(
                &context.repository,
                args(&["cat-file", "commit", oid]),
                token,
            )
            .await?;
        if bytes.len() > 128 * 1024 {
            return Err(AppError::BadRequest(
                "Commit metadata exceeds the 128 KiB preview limit".into(),
            ));
        }
        let separator = bytes
            .windows(2)
            .position(|pair| pair == b"\n\n")
            .ok_or_else(malformed)?;
        let header = &bytes[..separator];
        let mut parents = Vec::new();
        let mut author = None;
        for line in header.split(|byte| *byte == b'\n') {
            if let Some(parent) = line.strip_prefix(b"parent ") {
                let parent = std::str::from_utf8(parent).map_err(|_| malformed())?;
                require_oid(parent)?;
                parents.push(parent.into());
            } else if let Some(value) = line.strip_prefix(b"author ")
                && author.replace(value.to_vec()).is_some()
            {
                return Err(malformed());
            }
        }
        let author_header = author.ok_or_else(malformed)?;
        let author = String::from_utf8_lossy(&author_header);
        let mut fields = author.rsplitn(3, ' ');
        let _timezone = fields.next().ok_or_else(malformed)?;
        let seconds = fields
            .next()
            .ok_or_else(malformed)?
            .parse::<i64>()
            .map_err(|_| malformed())?;
        let identity = fields.next().ok_or_else(malformed)?;
        let split = identity.rfind('<').ok_or_else(malformed)?;
        let author_name = identity[..split].trim_end().to_owned();
        let author_email = identity[split + 1..]
            .strip_suffix('>')
            .ok_or_else(malformed)?
            .to_owned();
        let author_time = chrono::DateTime::from_timestamp(seconds, 0)
            .ok_or_else(malformed)?
            .to_rfc3339_opts(chrono::SecondsFormat::Secs, true);
        let subject = String::from_utf8_lossy(&bytes[separator + 2..])
            .lines()
            .next()
            .unwrap_or("")
            .chars()
            .take(4096)
            .collect();
        Ok(CommitData {
            parents,
            subject,
            author_header,
            author_name,
            author_email,
            author_time,
        })
    }
}

fn pick_parent(
    parents: &[String],
    mainline: Option<u32>,
    required: bool,
) -> AppResult<Option<&str>> {
    if parents.len() > 1 {
        match mainline {
            Some(index) if index > 0 && index as usize <= parents.len() => {
                Ok(Some(&parents[index as usize - 1]))
            }
            None if !required => Ok(None),
            _ => Err(AppError::BadRequest(
                "A merge commit requires an explicitly selected existing mainline parent".into(),
            )),
        }
    } else if mainline.is_some() {
        Err(AppError::BadRequest(
            "Mainline is only used when picking a merge commit".into(),
        ))
    } else {
        Ok(parents.first().map(String::as_str))
    }
}
fn pick_diff(oid: &str, parent: Option<&str>, patch: bool) -> Vec<String> {
    let mut command = if let Some(parent) = parent {
        args(&[
            "diff",
            "--no-ext-diff",
            "--no-textconv",
            "--no-renames",
            parent,
            oid,
        ])
    } else {
        args(&[
            "diff-tree",
            "--root",
            "-r",
            "--no-commit-id",
            "--no-ext-diff",
            "--no-textconv",
            "--no-renames",
            oid,
        ])
    };
    if patch {
        command.extend(args(&["--binary", "-p"]));
    } else {
        command.extend(args(&["--name-only", "-z"]));
    }
    command.push("--".into());
    command
}

File: apps/desktop/src/core/git/repository/cherrypick.ts
import {
  badRequest,
  conflict,
  malformed,
  nowRfc3339,
  oneLine,
  requireOid,
} from "../support";
import {
  gitIntegration,
  integrationSnapshot,
  ownsIntegration,
} from "./integration";
import { safely } from "./merge";
import { splitNul } from "./parse";
import {
  type Operation,
  RepositoryService,
  type RepositoryContext,
  commandError,
} from "./service";
import { protectLocalPaths, sameHead } from "./stash";
import type {
  CherryPickPreview,
  ExpectedState,
  RepositoryAction,
} from "./types";

/**
 * Cherry-pick and revert: the same owned single-commit sequence with the patch
 * applied in opposite directions.
 *
 * A port of `apps/runtime/src/git/repository/integration/cherry_pick.rs`. They
 * share every precondition, every ownership binding and every recovery path,
 * so they share an implementation; what differs is one Git verb and one
 * verification.
 */

interface CommitData {
  readonly parents: string[];
  readonly subject: string;
  /** The raw `author …` header line, compared byte for byte after a pick. */
  readonly authorHeader: string;
  readonly authorName: string;
  readonly authorEmail: string;
  readonly authorTime: string;
}

export async function cherryPickPreview(
  service: RepositoryService,
  root: string,
  requested: string,
  rawOid: string,
  mainline: number | undefined,
): Promise<CherryPickPreview> {
  const oid = rawOid.toLowerCase();
  requireOid(oid);
  const context = await service.context(root, requested);
  const commit = await pickCommit(service, context, oid);
  const parent = pickParent(commit.parents, mainline, false);
  const patch =
    commit.parents.length > 1 && mainline === undefined
      ? null
      : (
          await service.read(context.repository, pickDiff(oid, parent, true))
        ).toString("utf8");
  return {
    targetOid: oid,
    parents: commit.parents,
    subject: commit.subject,
    authorName: commit.authorName,
    authorEmail: commit.authorEmail,
    authorTime: commit.authorTime,
    mainline: mainline ?? null,
    patch,
  };
}

export async function startCherryPick(
  service: RepositoryService,
  context: RepositoryContext,
  action: RepositoryAction,
  expected: ExpectedState,
  operation: Operation,
): Promise<void> {
  let targetOid: string;
  let mainline: number | null;
  let recordOrigin = false;
  let expectedStateToken: string;
  let reverting: boolean;
  if (action.kind === "startCherryPick") {
    targetOid = action.targetOid;
    mainline = action.mainline;
    recordOrigin = action.recordOrigin;
    expectedStateToken = action.expectedStateToken;
    reverting = false;
  } else if (action.kind === "revert") {
    targetOid = action.targetOid;
    mainline = action.mainline;
    expectedStateToken = action.expectedStateToken;
    reverting = true;
  } else {
    throw malformed();
  }
  targetOid = targetOid.toLowerCase();
  const signal = operation.controller.signal;
  const state = await integrationSnapshot(service, context, signal);
  if (
    state.kind !== "none" ||
    !sameHead(state.head, expected) ||
    state.stateToken !== expectedStateToken
  ) {
    throw conflict(
      "Git state changed or an integration is already active; refresh and confirm again",
    );
  }
  if (
    state.dirty ||
    state.conflicts.length > 0 ||
    state.head.headOid === null ||
    state.head.branch === null
  ) {
    throw conflict(
      reverting
        ? "Revert requires a clean index and worktree on a committed local branch"
        : "Cherry-pick requires a clean index and worktree on a committed local branch",
    );
  }
  const commit = await pickCommit(service, context, targetOid, signal);
  const parent = pickParent(commit.parents, mainline ?? undefined, true);
  const paths = await service.read(
    context.repository,
    pickDiff(targetOid, parent, false),
    signal,
  );
  const touched = new Set(
    splitNul(paths)
      .filter((path) => path.length > 0)
      .map((path) => path.toString("utf8")),
  );
  await protectLocalPaths(service, context, touched, signal);

  const kind = reverting ? "revert" : "cherryPick";
  service.integrations.set(context.repository, {
    sessionId: operation.snapshot.id,
    kind,
    mainline,
    original: expected,
    targetOid,
    marker: null,
  });
  const command = [
    "-c",
    "core.editor=:",
    "-c",
    "rerere.enabled=false",
    reverting ? "revert" : "cherry-pick",
    "--no-rerere-autoupdate",
  ];
  // The recorded message is Git's own; nothing here opens an editor.
  if (reverting) command.push("--no-edit");
  if (mainline !== null) command.push("--mainline", String(mainline));
  if (recordOrigin) command.push("-x");
  command.push("--", targetOid);

  let output;
  let failure: unknown;
  try {
    output = await service.output(
      context.repository,
      command,
      service.commandTimeoutMs,
      signal,
      {
        onMutationStarted: () => {
          operation.mutationStarted = true;
        },
      },
    );
  } catch (error) {
    failure = error;
  }
  const actual = await safely(() => gitIntegration(service, context));
  const head = await safely(() => service.head(context.repository));
  let owned = false;
  if (actual.ok && head.ok) {
    const owner = service.integrations.get(context.repository);
    if (owner !== undefined) {
      if (
        actual.value.kind === kind &&
        actual.value.targetOid === targetOid &&
        sameHead(head.value, expected)
      ) {
        owner.marker = actual.value.marker;
        owned = ownsIntegration(owner, actual.value, head.value);
      } else if (actual.value.kind === "none") {
        service.integrations.delete(context.repository);
      }
    }
  }
  if (failure !== undefined) throw failure;
  if (!actual.ok) throw actual.error;
  if (!head.ok) throw head.error;
  if (output === undefined) throw malformed();
  if (owned && (output.status === 0 || output.status === 1)) {
    operation.awaitingResolution = true;
    return;
  }
  if (output.status !== 0) throw commandError(output);
  if (actual.value.kind !== "none") {
    throw conflict(
      `${kind} left an unverified Git sequence; inspect it before another action`,
    );
  }
  if (reverting) {
    await verifyRevert(
      service,
      context,
      targetOid,
      expected,
      head.value,
      signal,
    );
  } else {
    await verifyCherryPick(
      service,
      context,
      targetOid,
      expected,
      head.value,
      signal,
    );
  }
}

/**
 * A finished revert is one new commit on the same branch whose parent is the
 * confirmed original head, and whose tree undoes the target's patch. The author
 * is the current user, so nothing is checked against the source.
 */
export async function verifyRevert(
  service: RepositoryService,
  context: RepositoryContext,
  target: string,
  expected: ExpectedState,
  head: ExpectedState,
  signal?: AbortSignal,
): Promise<void> {
  const original = expected.headOid;
  const current = head.headOid;
  if (original === null || current === null) throw malformed();
  const produced = await pickCommit(service, context, current, signal);
  if (
    head.branch !== expected.branch ||
    current === original ||
    produced.parents.length !== 1 ||
    produced.parents[0] !== original
  ) {
    throw conflict(
      "The resulting revert commit does not sit on the confirmed parent and branch; inspect HEAD before another action",
    );
  }
  // The target must still be an ancestor: a revert that no longer relates to
  // the commit it claims to undo is not a result we accept.
  const reachable = await service.output(
    context.repository,
    ["merge-base", "--is-ancestor", target, current],
    15_000,
    signal,
  );
  if (reachable.status !== 0) {
    throw conflict(
      "The reverted commit is not reachable from the new HEAD; inspect the repository",
    );
  }
}

export async function verifyCherryPick(
  service: RepositoryService,
  context: RepositoryContext,
  target: string,
  expected: ExpectedState,
  head: ExpectedState,
  signal?: AbortSignal,
): Promise<void> {
  const original = expected.headOid;
  const current = head.headOid;
  if (original === null || current === null) throw malformed();
  const source = await pickCommit(service, context, target, signal);
  const produced = await pickCommit(service, context, current, signal);
  if (
    head.branch !== expected.branch ||
    current === original ||
    produced.parents.length !== 1 ||
    produced.parents[0] !== original ||
    produced.authorHeader !== source.authorHeader
  ) {
    throw conflict(
      "Resulting cherry-pick commit does not match its confirmed parent and source author; inspect HEAD before another action",
    );
  }
}

/**
 * The commit header, parsed rather than walked.
 *
 * `cat-file commit` is read instead of `git log` so shallow boundary rewriting
 * cannot misrepresent a non-root commit as a root.
 */
async function pickCommit(
  service: RepositoryService,
  context: RepositoryContext,
  oid: string,
  signal?: AbortSignal,
): Promise<CommitData> {
  requireOid(oid);
  const kind = oneLine(
    await service.read(context.repository, ["cat-file", "-t", oid], signal),
  );
  if (kind !== "commit") {
    throw badRequest(
      "Cherry-pick target must be a commit object ID, not a tag or tree",
    );
  }
  const bytes = await service.read(
    context.repository,
    ["cat-file", "commit", oid],
    signal,
  );
  if (bytes.length > 128 * 1024) {
    throw badRequest("Commit metadata exceeds the 128 KiB preview limit");
  }
  const separator = bytes.indexOf("\n\n");
  if (separator < 0) throw malformed();
  const header = bytes.subarray(0, separator).toString("utf8");
  const parents: string[] = [];
  let authorHeader: string | undefined;
  for (const line of header.split("\n")) {
    if (line.startsWith("parent ")) {
      const parent = line.slice("parent ".length);
      requireOid(parent);
      parents.push(parent);
    } else if (line.startsWith("author ")) {
      if (authorHeader !== undefined) throw malformed();
      authorHeader = line.slice("author ".length);
    }
  }
  if (authorHeader === undefined) throw malformed();
  const lastSpace = authorHeader.lastIndexOf(" ");
  const secondLast = authorHeader.lastIndexOf(" ", lastSpace - 1);
  if (lastSpace < 0 || secondLast < 0) throw malformed();
  const seconds = Number.parseInt(
    authorHeader.slice(secondLast + 1, lastSpace),
    10,
  );
  if (Number.isNaN(seconds)) throw malformed();
  const identity = authorHeader.slice(0, secondLast);
  const open = identity.lastIndexOf("<");
  if (open < 0 || !identity.endsWith(">")) throw malformed();
  const authorName = identity.slice(0, open).trimEnd();
  const authorEmail = identity.slice(open + 1, -1);
  const authorTime = `${new Date(seconds * 1000).toISOString().slice(0, 19)}Z`;
  const subject = [
    ...(bytes
      .subarray(separator + 2)
      .toString("utf8")
      .split("\n")[0] ?? ""),
  ]
    .slice(0, 4096)
    .join("");
  return {
    parents,
    subject,
    authorHeader,
    authorName,
    authorEmail,
    authorTime,
  };
}

function pickParent(
  parents: readonly string[],
  mainline: number | undefined,
  required: boolean,
): string | undefined {
  if (parents.length > 1) {
    if (mainline !== undefined && mainline > 0 && mainline <= parents.length) {
      return parents[mainline - 1];
    }
    if (mainline === undefined && !required) return undefined;
    throw badRequest(
      "A merge commit requires an explicitly selected existing mainline parent",
    );
  }
  if (mainline !== undefined) {
    throw badRequest("Mainline is only used when picking a merge commit");
  }
  return parents[0];
}

function pickDiff(
  oid: string,
  parent: string | undefined,
  patch: boolean,
): string[] {
  const command =
    parent === undefined
      ? [
          "diff-tree",
          "--root",
          "-r",
          "--no-commit-id",
          "--no-ext-diff",
          "--no-textconv",
          "--no-renames",
          oid,
        ]
      : ["diff", "--no-ext-diff", "--no-textconv", "--no-renames", parent, oid];
  command.push(...(patch ? ["--binary", "-p"] : ["--name-only", "-z"]), "--");
  return command;
}

export { nowRfc3339 };
